### PR TITLE
Support business prepop survey launch

### PR DESCRIPTION
**Pull request by liamtoozer**
### What is the context of this PR?
We want to be able to launch business schemas that use supplementary data. This service has been altered to serve SDS metadata and unit data back to the caller. 

A similar pattern to cache the mock unit data has been used, as in our mock CIR API. Happy to change this pattern to something else, as there isn't much SDS mock data at the moment. We could simply preload the data at startup instead, but see what you think might be best for now.

For this PR, we now do the following:

* Load the mock data from the [SDS schema examples](https://github.com/ONSdigital/sds-schema-definitions/tree/main/examples) at startup
* Dynamically create metadata and deterministically generate a `dataset_id` for each metadata chunk (similar to the mock CIR repo for its data)
* Load unit data for each survey ID with [SDS schema examples](https://github.com/ONSdigital/sds-schema-definitions/tree/main/examples). This maps internally to the dataset IDs **Note:** Survey IDs are hardcoded so we can map them to the folder names that contain the unit data.
* Identifiers and period IDs are still unused at this point.

### How to review
1. Run `make load-mock-unit-data` to pull down the example unit data from the SDS schema repo
2. Run the API
3. Make some requests to the API endpoints, e.g.: `/v1/dataset_metadata?survey_id=014&period_id=1` and `/v1/unit_data?dataset_id=d8afa921-1305-d553-d2c6-955a6db2cc2d&identifier=1`. Ensure that the data returned is as expected
4. Checkout the [Launcher PR branch](https://github.com/ONSdigital/eq-questionnaire-launcher/pull/74) and try launching various business surveys that make use of supplementary data. Ensure that the test SDS schemas still load correctly as well.

Note: As mentioned [in the comments](https://github.com/ONSdigital/eq-runner-mock-sds/pull/11#pullrequestreview-1973332853), Prodcom and BRES/BRS schemas cannot currently be tested

[Jira ticket](https://jira.ons.gov.uk/browse/ECI-1122)

### CODE REVIEW

File: scripts/load_mock_unit_data.sh
Comment (berroar) [inline scripts/load_mock_unit_data.sh]: Difficult to test this part until the other PR merges as you mention, although the above seems to work and I am able to launch SDS schemas

File: app/main.py
Comment (berroar) [inline app/main.py:38]: I guess worth noting that prodcom and bres/brs cannot be tested yet as there are no schemas for these atm

File: app/main.py
Comment (berroar) [inline app/main.py:55]: I guess this could go out of sync with the SDS model if they make changes, is it possible/should we install this as a dependency?

File: app/main.py
Comment (berroar) [inline app/main.py:90]: The type hints in this file should use the types from the class rather than being set again?

File: app/main.py
Comment (berroar) [inline app/main.py:107]: Is there a reason why we return all the datasets here? Could we just pass in the survey_id and just return the one we need? Then we wouldn't need the nested loops?

File: app/main.py
Comment (berroar) [inline app/main.py:93]: We have `_, dataset_to_unit_data_map` here and `dataset_metadata_collection, _` for the other usage of this method. Do we need to generate both of these maps every time? Could either be a separate method or some conditional logic so it only generates what is needed?

File: Dockerfile
Comment (petechd) [inline Dockerfile:33]: You might have to go back to the root directory before running this make command.

File: app/main.py
Comment (liamtoozer) [reply app/main.py:107]: I think I tried to keep everything similar to how we do things with the mock CIR API, i.e. get all of the data upfront, cache it, and then filter what we need in return. As we're not dealing with much data in the mock SDS though, that approach might be a bit much, so think it makes sense to simplify it a bit more 👍

File: app/main.py
Comment (liamtoozer) [reply app/main.py:55]: That's a good point 👍  Would we need to consider that the `.model_validate()` calls for each of the models might break at runtime, as we're configuring the data before that call? I think if any key/vals are different to what's expected, then we get a validation error? It might be okay though, so happy to go with whatever might be best!

File: scripts/load_mock_unit_data.sh
Comment (liamtoozer) [reply scripts/load_mock_unit_data.sh]: I should have popped the PR-convenience chunk into a separate file instead to make it a bit easier for review, and not muddy the water too much. Will do this now, with a reminder to revert before merge, thanks 👍

File: scripts/pr_convenience.sh
Comment (liamtoozer) [inline scripts/pr_convenience.sh]: Reminder: Remove this script before merge

File: Dockerfile
Comment (liamtoozer) [reply Dockerfile:33]: I had to add a dependency to get the `make` command to work so it should be okay now 🤞 Thanks for the heads up. The files loaded from the dependent SDS schema repo will be from `main` branch, so will not be in a testable state at the moment until that PR on that repo is merged, but it should build okay hopefully

File: app/main.py
Comment (liamtoozer) [reply app/main.py:55]: After a discussion in a tech session, it's been recommended to copy these classes in for now. In future, if SDS were to host these classes centrally and publish them, we could then import them as a dependency

File: app/main.py
Comment (liamtoozer) [reply app/main.py:38]: Updated the PR description to explain, but for clarity, should I also pop this in as an inline comment too?

File: app/main.py
Comment (liamtoozer) [reply app/main.py:90]: I've tried using dot notation e.g. `dataset_id: UnitData.dataset_id`, but pydantic gives an error `AttributeError: dataset_id` when I try to start up the service. I'm not certain, but it might be trying to access the class member value rather than the type of that member? There might be a way around this but I'm not sure?

File: app/main.py
Comment (petechd) [reply app/main.py:93]: I agree, this is ineffective, calling this method twice and using only part of what is returned, we want to have two methods, more specific ones, that return only what's needed.

Comment (liamtoozer) [reply app/main.py:93]: I might have misunderstood the questions sorry, but with the LRU caching decorator on `load_mock_data()`, when either `/v1/unit_data` or `/v1/dataset_metadata` end point is first called, the `load_mock_data()` will cache the result for both of the mappings. Subsequent calls to `load_mock_data()` will reuse this cached result for both mappings. So the mappings are only generated once for the lifecycle of the mock SDS instance. 

I've checked the object ID references for both maps each time `load_mock_data()` is called, and they remain the same, so it might be okay if we then filter the result afterwards? This is the same pattern used in the mock CIR as well (although that was for a much larger scope of data as it was loading in all of the schemas)

File: scripts/load_mock_unit_data.sh
Comment (berroar) [inline scripts/load_mock_unit_data.sh]: Only minor but do we need this as a var since it's only used once?

File: app/main.py
Comment (berroar) [reply app/main.py:90]: Guessing there was no way around this?

Comment (liamtoozer) [reply app/main.py:90]: I've tried a few different things but I can't seem to get around it. It keeps giving me the `AttributeError: dataset_id` on startup when I try changing it to `dataset_id: UnitData.dataset_id`

File: app/main.py
Comment (berroar) [reply app/main.py:107]: As discussed won't hold this up since it's a mock and we're copying the CIR pattern 👍

**Review by berroar — APPROVED** [review]

**Review by petechd — APPROVED** [review]

**Pull request merged**
